lightning/src/ln/channelmanager.rs

```diff
@@ -7906,6 +7704,351 @@ where
+	).unwrap().secret_bytes();
+
+	if msg.onion_routing_packet.version != 0 {
+		//TODO: Spec doesn't indicate if we should only hash hop_data here (and in other
+		//sha256_of_onion error data packets), or the entire onion_routing_packet. Either way,
+		//the hash doesn't really serve any purpose - in the case of hashing all data, the
+		//receiving node would have to brute force to figure out which version was put in the
+		//packet by the node that send us the message, in the case of hashing the hop_data, the
+		//node knows the HMAC matched, so they already know what is there...
+		return_malformed_err!("Unknown onion packet version", 0x8000 | 0x4000 | 4);
+	}
+	macro_rules! return_err {
+		($msg: expr, $err_code: expr, $data: expr) => {
+			{
+				log_info!(logger, "Failed to accept/forward incoming HTLC: {}", $msg);
+				return Err(HTLCFailureMsg::Relay(msgs::UpdateFailHTLC {
+					channel_id: msg.channel_id,
+					htlc_id: msg.htlc_id,
+					reason: HTLCFailReason::reason($err_code, $data.to_vec())
+						.get_encrypted_failure_packet(&shared_secret, &None),
+				}));
+			}
+		}
+	}
+
+	let next_hop = match onion_utils::decode_next_payment_hop(
+		shared_secret, &msg.onion_routing_packet.hop_data[..], msg.onion_routing_packet.hmac,
+		msg.payment_hash, node_signer
+	) {
+		Ok(res) => res,
+		Err(onion_utils::OnionDecodeErr::Malformed { err_msg, err_code }) => {
+			return_malformed_err!(err_msg, err_code);
+		},
+		Err(onion_utils::OnionDecodeErr::Relay { err_msg, err_code }) => {
+			return_err!(err_msg, err_code, &[0; 0]);
+		},
+	};
+
+	let next_packet_details = match next_hop {
+		onion_utils::Hop::Forward {
+			next_hop_data: msgs::InboundOnionPayload::Forward {
+				short_channel_id, amt_to_forward, outgoing_cltv_value
+			}, ..
+		} => {
+			let next_packet_pubkey = onion_utils::next_hop_pubkey(secp_ctx,
+				msg.onion_routing_packet.public_key.unwrap(), &shared_secret);
+			NextPacketDetails {
+				next_packet_pubkey, outgoing_scid: short_channel_id,
+				outgoing_amt_msat: amt_to_forward, outgoing_cltv_value
+			}
+		},
+		onion_utils::Hop::Receive { .. } => return Ok((next_hop, shared_secret, None)),
+		onion_utils::Hop::Forward { next_hop_data: msgs::InboundOnionPayload::Receive { .. }, .. } |
+			onion_utils::Hop::Forward { next_hop_data: msgs::InboundOnionPayload::BlindedReceive { .. }, .. } =>
+		{
+			return_err!("Final Node OnionHopData provided for us as an intermediary node", 0x4000 | 22, &[0; 0]);
+		}
+	};
+
+	Ok((next_hop, shared_secret, Some(next_packet_details)))
+}
+
+fn check_incoming_htlc_cltv(
+	cur_height: u32, outgoing_cltv_value: u32, cltv_expiry: u32
+) -> Result<(), (&'static str, u16, bool)> {
+	if (cltv_expiry as u64) < (outgoing_cltv_value) as u64 + MIN_CLTV_EXPIRY_DELTA as u64 {
+		// We really should set `incorrect_cltv_expiry` here but as we're not
+		// forwarding over a real channel we can't generate a channel_update
+		// for it. Instead we just return a generic temporary_node_failure.
```

Comment:
Rather than keeping this behavior, we should always return the correct error in this method and convert it if needed at the callsite:

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index e9433a199..c4a37c80a 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -3030,10 +3030,16 @@ where

                        let cur_height = self.best_block.read().unwrap().height() + 1;

-                       if let Err((err_msg, code, include_chan_update_opt)) = check_incoming_htlc_cltv(
+                       if let Err((err_msg, code)) = check_incoming_htlc_cltv(
                                cur_height, outgoing_cltv_value, msg.cltv_expiry
                        ) {
-                               let chan_update_opt = if include_chan_update_opt { chan_update_opt } else { None };
+                               if code & 0x1000 != 0 && chan_update_opt.is_none() {
+                                       // We really should set `incorrect_cltv_expiry` here but as we're not
+                                       // forwarding over a real channel we can't generate a channel_update
+                                       // for it. Instead we just return a generic temporary_node_failure.
+                                       break Some((err_msg, 0x2000 | 2, None))
+                               }
+                               let chan_update_opt = if code & 0x1000 != 0 { chan_update_opt } else { None };
                                break Some((err_msg, code, chan_update_opt));
                        }

@@ -8015,14 +8021,11 @@ where

 fn check_incoming_htlc_cltv(
        cur_height: u32, outgoing_cltv_value: u32, cltv_expiry: u32
-) -> Result<(), (&'static str, u16, bool)> {
+) -> Result<(), (&'static str, u16)> {
        if (cltv_expiry as u64) < (outgoing_cltv_value) as u64 + MIN_CLTV_EXPIRY_DELTA as u64 {
-               // We really should set `incorrect_cltv_expiry` here but as we're not
-               // forwarding over a real channel we can't generate a channel_update
-               // for it. Instead we just return a generic temporary_node_failure.
                return Err((
                                "Forwarding node has tampered with the intended HTLC values or origin node has an obsolete cltv_expiry_delta",
-                               0x2000 | 2, false,
+                               0x1000 | 13, // incorrect_cltv_expiry
                ));
        }
        // Theoretically, channel counterparty shouldn't send us a HTLC expiring now,
```

Note that we also don't need to return whether a channel update is needed, we can detect it based on whether the error code includes `0x1000`.

Reply:
cool